streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java

```diff
@@ -364,26 +365,24 @@ public static void waitForCompletion(final KafkaStreams streams,
                                                                                     final long waitTime) throws InterruptedException {
         final List<KeyValue<K, V>> accumData = new ArrayList<>();
         try (final Consumer<K, V> consumer = createConsumer(consumerConfig)) {
-            final TestCondition valuesRead = new TestCondition() {
-                @Override
-                public boolean conditionMet() {
-                    final List<KeyValue<K, V>> readData =
-                            readKeyValues(topic, consumer, waitTime, expectedRecords.size());
-                    accumData.addAll(readData);
-
-                    final Map<K, V> finalData = new HashMap<>();
+            final TestCondition valuesRead = () -> {
+                final List<KeyValue<K, V>> readData =
+                        readKeyValues(topic, consumer, waitTime, expectedRecords.size());
+                accumData.addAll(readData);
 
-                    for (final KeyValue<K, V> keyValue : accumData) {
-                        finalData.put(keyValue.key, keyValue.value);
-                    }
+                final Map<K, List<V>> finalData = new HashMap<>();
 
-                    for (final KeyValue<K, V> keyValue : expectedRecords) {
-                        if (!keyValue.value.equals(finalData.get(keyValue.key)))
-                            return false;
-                    }
+                for (final KeyValue<K, V> keyValue : accumData) {
+                    final List<V> keyData = finalData.computeIfAbsent(keyValue.key, k -> new ArrayList<>());
```

Comment:
Updated here to support having results with the same key and different results, this was causing a test failure when the expected test results were something like `A:value`, `A:value_2` etc.

Reply:
Why `computeIfAbsent`? I though later key-value pairs should overwrite previous? Isn't this "keep first", but we want "keep last"?

Reply:
yes,  but that's the problem which why it's a key -> list.  I have a join in the integration test (https://github.com/apache/kafka/pull/5517/files#diff-21c24c91023f38c26b614af254ab50f2R206) returning `A, value 1`, `A, value 2`, and `A, value 3` but fails because only `A, value 3` is returned due to overwriting of the previous key.

Reply:
We need to check for ordering for the same key as well: if the expected results is `{A, 1}, {A, 2}` but the consumer returns `{A, 2}, {A, 1}`, this test will still pass right?

Reply:
Yes, that's correct. But the test asserts that the returned results list equals the expected data out of order will cause the test to fail.

Reply:
Yes, and more generally, my understanding is that we also need to consider if the consumer may return a "super-list" of the expected values, for example (still assume the above expected `{A, 1}, {A, 2}`):

* `{A, 1}, {A, 2}`: this should be correct.
* `{A, 1}, {A, 1.5}, {A, 2}`: this should be correct.
* `{A, 1}, {A, 2}, {A, 2.5}`: this should be wrong, since `{A, 2}` should be the last for the key.
* `{A, 2}, {A, 1}`: this should be wrong.
* `{A, 2}, {A, 2.5}, {A, 1}`: this should be wrong.

Reply:
My bad. `computeIfAbsent(keyValue.key, k -> new ArrayList<>());` computes a new `ArrayList` when key is missing. Was skipping over the code too quickly.